yang/vyatta-protocols-frr-isis-v1.yang

```diff
@@ -0,0 +1,581 @@
+      }
+      enum "level-1-2" {
+        value 3;
+        configd:help "This enum indicates capability for both levels.";
+        description "This enum indicates capability for both levels.";
+      }
+    }
+    configd:help "This type defines IS-IS level of an object.";
+    description "This type defines IS-IS level of an object.";
+  } 
+
+  typedef network-type {
+    type enumeration {
+      enum "unknown" {
+        value 0;
+        configd:help "Unknown network type. Only valid as a state.";
+        description "Unknown network type. Only valid as a state.";
+      }
+      enum "broadcast" {
+        value 1;
+        configd:help "Broadcast circuit network-type.";
+        description "Broadcast circuit network-type.";
+      }
+      enum "point-to-point" {
+        value 2;
+        configd:help "Point-to-point circuit network-type."; 
+        description "Point-to-point circuit network-type.";
+      }
+      enum "loopback" {
+        value 3;
+        configd:help "Loopback circuit network-type. Only valid as a state.";
+        description "Loopback circuit network-type. Only valid as a state.";
+      }
+    }
+  }
+
+  typedef snpa {
+    type string {
+      length "0 .. 20";
+    }
+    configd:help "This type defines the Subnetwork Point of Attachment (SNPA) format.
+    The SNPA should be encoded according to the rules specified for the particular type
+    of subnetwork being used. As an example, for an ethernet subnetwork, the SNPA is 
+    encoded as a MAC address like '00aa.bbcc.ddee'.";
+
+    description "This type defines the Subnetwork Point of Attachment (SNPA) format.
+    The SNPA should be encoded according to the rules specified for the particular type
+    of subnetwork being used. As an example, for an ethernet subnetwork, the SNPA is
+    encoded as a MAC address like '00aa.bbcc.ddee'.";
+  }
+
+  typedef net-address {
+    type string {
+      pattern "[a-fA-F0-9]{2}(\\.[a-fA-F0-9]{4}){3,9}\\.[a-fA-F0-9]{2}";
+    }
+    configd:help "This type defines an OSI NET address using a pattern,
+    An example net-address is 49.0123.6452.1972.00";
+    description "This type defines an OSI NET address using a pattern,
+    An example net-address is 49.0123.6452.1972.00";
+  }
+
+  typedef metric-style-type {
+    type enumeration {
+      enum "narrow" {
+        value 0;
+        configd:help "This enum describes narrow metric style";
+        description "This enum describes narrow metric style";
+        reference "RFC1195";
+      }
+      enum "wide" {
+        value 1;
+        configd:help "This enum describes wide metric style";
+        description "This enum describes wide metric style";
+        reference "RFC5305";
+      }
+      enum "transition" {
+        value 2;
+        configd:help "This enum describes transition metric style";
+        description "This enum describes transition metric style";
+      }
+    }
+  }
+
+  grouping isis-password {
+    choice type-choice {
```

Comment:
DANOS does not yet support choice - it will compile, but is ignored.  So, this isn't going to work.  It also raises the question of what testing has been done on this?

Reply:
Hi Wivory, 

We have tested "choice" and found it to be working before sending for review. 
We have attached 5 screenshot of the following below. 
1. Screenshot of the DANOS CLI. 
2. Screenshot of the frr.json file from location:- /etc/vyatta-routing/frr.json 
3. Screenshot of the frr.conf file from location:- /etc/vyatta-routing/frr.conf
4. And finally from the frr.conf file from location:- /etc/frr/frr.conf
5. Screenshot of the set command from DANOS CLI.

Are you suggesting any other way to validate that it is working ? 

Thanks, 

![1_DANOS_cli](https://user-images.githubusercontent.com/62253226/78272878-b5102880-752b-11ea-87ce-85b29695aaea.png)
![2_DANOS_vyatta_routing_frr_json](https://user-images.githubusercontent.com/62253226/78273073-f6a0d380-752b-11ea-9213-a30a5258de88.png)
![3_DANOS_vyatta_routing_frr_conf](https://user-images.githubusercontent.com/62253226/78273102-fd2f4b00-752b-11ea-9fe9-034050c80bef.png)
![4_DANOS_etc_frr_conf](https://user-images.githubusercontent.com/62253226/78273112-028c9580-752c-11ea-8eb0-bd29bb376cb7.png)
![5_DANOS_isis_cli_password](https://user-images.githubusercontent.com/62253226/78273132-09b3a380-752c-11ea-8968-fc74b6d616d8.png)

Reply:
Sorry - perhaps I wasn't clear what I meant by 'ignored'.  We ignore the 'choice' element, and allow all choices to be configured as shown here:
`vyatta@vm-at-1# set protocols isis 1 area-password clear foo
[edit]
vyatta@vm-at-1# show protocols isis
+isis 1 {
+       area-password {
+               clear foo
+       }
+}
[edit]
vyatta@vm-at-1# set protocols isis 1 area-password md5 bar
[edit]
vyatta@vm-at-1# show protocols isis
+isis 1 {
+       area-password {
+               clear foo
+               md5 bar
+       }
+}
`
If we had implemented choice correctly, setting the md5 option would have removed the clear option.  You certainly couldn't commit both options with choice implemented correctly - but you will find that on DANOS currently, you can.

So, I guess you can keep the YANG as-is, but for now you will be able to configure both 'clear' and 'md5' at the same time.

Reply:
Okay, 
Thanks.

Reply:
Actually you must not keep this yang or there is a risk that someone configures both options, and then when we fix choice, the config becomes invalid, ie this would be backwards incompatible.

Reply:
Hi Wivory, 

We have found a way which does not use "choice" and hence it should be good. 
We are committing this change now.
 
grouping isis-password {
  leaf clear {
    configd:help "Clear-text authentication type.";
    description "Clear-text authentication type.";
    type string {
      length "1..254";
    } 
    must "not(../md5)" {
      error-message "Can't configure both clear and md5 for password";
    }
  }
  leaf md5 {
    configd:help "MD5 authentication type.";
    description "MD5 authentication type.";
    type string {
      length "1..254";
    } 
    must "not(../clear)" {
      error-message "Can't configure both clear and md5 for password";
    } 
  } 
}

Reply:
Yes, that will do the trick.  You only need it on one of the 2 leaves though - no need to check both ways.

Reply:
Okay, we rectified it and committed. 

Thanks.